Bottom: 869637c8d50d60612883f939db61fbef86733ac0
Top:    ad711dcbfa681503c16b448c5dc94d9ee4d24ff2
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-03-15 13:54:36 +0800

Refresh of virtio-net

---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 2f762f2..7509cd4 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -356,9 +356,11 @@ static void virtnet_rx_recycle(struct receive_queue *rq, struct page *page)
 {
 	struct virtnet_page_cache *cache = &rq->page_cache;
 
-	if (page_ref_count(page) == 2 && cache->index < VIRTNET_CACHE_SIZE)
-		cache->pages[cache->index++] = page;
-	put_page(page);
+	if (page_ref_count(page) == 1 && cache->index < VIRTNET_CACHE_SIZE) {
+		if (rq->alloc_frag.page != page)
+			cache->pages[cache->index++] = page;
+	} else
+		put_page(page);
 }
 
 static bool virtnet_xdp_xmit(struct virtnet_info *vi,
